src/pool/srv_pool_scrub.c

```diff
@@ -4,324 +4,771 @@
 {
-	char *sec = getenv("DAOS_CSUM_SCRUB_PAUSE_SEC");
+	ctx->sc_cur_oid = param->ip_oid;
+	ctx->sc_dkey = param->ip_dkey;
+	ctx->sc_epoch = entry->ie_epoch;
+	ctx->sc_minor_epoch = entry->ie_minor_epc;
 
-	return sec != NULL ? atoll(sec) : 24 * 60 * 60;
+	ctx->sc_iod.iod_size = entry->ie_rsize;
+	ctx->sc_iod.iod_nr = 1;
+	ctx->sc_iod.iod_type = type == VOS_ITER_RECX ? DAOS_IOD_ARRAY :
+			       DAOS_IOD_SINGLE;
+	ctx->sc_iod.iod_name = param->ip_akey;
+	ctx->sc_iod.iod_recxs = &entry->ie_recx;
+
+	ctx->sc_csum_to_verify = &entry->ie_csum;
+
+	ctx->sc_vos_iter_handle = ih;
+}
+
+static bool
+oids_are_same(daos_unit_oid_t a, daos_unit_oid_t b)
+{
+	return daos_unit_oid_compare(a, b) == 0;
+}
+
+static bool
+keys_are_same(daos_key_t key1, daos_key_t key2)
+{
+	if (key1.iov_len != key2.iov_len)
+		return 0;
+
+	return memcmp(key1.iov_buf, key2.iov_buf, key1.iov_len) == 0;
+}
+
+static bool
+epoch_is_same(daos_epoch_t a, daos_epoch_t b)
+{
+	return a == b;
 }
 
 /** vos_iter_cb_t */
 static int
-obj_iter_scrub_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+obj_iter_scrub_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
 		  vos_iter_type_t type, vos_iter_param_t *param,
 		  void *cb_arg, unsigned int *acts)
 {
 	struct scrub_ctx	*ctx = cb_arg;
-	struct daos_csummer	*csummer = ctx->cur_cont->sc_csummer;
+	int			 rc;
 
-	if (!(type == VOS_ITER_RECX || type == VOS_ITER_SINGLE))
+	if (ctx->sc_pool->sp_scrub_sched == DAOS_SCRUB_SCHED_OFF) {
+		C_TRACE("scrubbing is off now, aborting ...");
+		*acts |= VOS_ITER_CB_ABORT;
 		return 0;
+	}
 
-	C_TRACE("Scrubbing akey: "DF_KEY", type: %s, record size: "
-			DF_U64", extent: "DF_RECX", csum_size: %d\n",
-		DP_KEY(&param->ip_akey),
-		(type == VOS_ITER_RECX) ? "ARRAY" : "SV",
-		entry->ie_rsize,
-		DP_RECX(entry->ie_orig_recx),
-		daos_csummer_get_csum_len(csummer)
-	);
-
-	/** TODO - implement the actual scrubbing of the checksum by fetching
-	 * the data, calculating a new checksum, and comparing to the stored
-	 * checksum.
-	 */
-	ctx->pool_csums_scrubbed++;
+	switch (type) {
+	case VOS_ITER_OBJ:
+		if (oids_are_same(ctx->sc_cur_oid, entry->ie_oid)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_cur_oid, 0, sizeof(ctx->sc_cur_oid));
+		} else {
+			ctx->sc_cur_oid = entry->ie_oid;
+		}
+		break;
+	case VOS_ITER_DKEY:
+		if (keys_are_same(ctx->sc_dkey, entry->ie_key)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_dkey, 0, sizeof(ctx->sc_dkey));
+		} else {
+			ctx->sc_dkey = param->ip_dkey;
+		}
+		break;
+	case VOS_ITER_AKEY:
+		if (keys_are_same(ctx->sc_iod.iod_name, entry->ie_key)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_iod, 0, sizeof(ctx->sc_iod));
+		} else {
+			ctx->sc_iod.iod_name = param->ip_akey;
+		}
+		break;
+	case VOS_ITER_SINGLE:
+	case VOS_ITER_RECX: {
+		if (epoch_is_same(ctx->sc_epoch, entry->ie_epoch) &&
+			epoch_is_same(ctx->sc_minor_epoch,
+				      entry->ie_minor_epc)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			ctx->sc_epoch = 0;
+			ctx->sc_minor_epoch = 0;
```

Comment:
I don't remember why I put VOS_ITER_CB_SKIP here. You're right, it might not be needed. I'll test to see if it works without. Yes, sc_epoch and sc_minor_epoch are used to identify SV/EV records, and while two disjoint EV tree records may have same epoch/minor_epoch, my understanding is that the entry at this point shouldn't be of a disjoint EV tree record. Shouldn't it be of the same EV tree?

Reply:
Same EV tree may have two disjoint records like [0, 100)@epoch 1 and [200, 300)@epoch1 , right?

Reply:
Yep, you're right. I fixed it so it also looks at the recx of the entry.